mesg/deploy_deployer.go

```diff
@@ -0,0 +1,135 @@
+package mesg
+
+import (
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/url"
+	"os"
+
+	"github.com/docker/docker/pkg/archive"
+	"github.com/logrusorgru/aurora"
+	"github.com/mesg-foundation/core/database/services"
+	"github.com/mesg-foundation/core/service"
+	"github.com/mesg-foundation/core/service/importer"
+	uuid "github.com/satori/go.uuid"
+	git "gopkg.in/src-d/go-git.v4"
+	"gopkg.in/src-d/go-git.v4/plumbing"
+)
+
+type serviceDeployer struct {
+	Statuses chan string
+
+	mesg *MESG
+}
+
+func newServiceDeployer(mesg *MESG) *serviceDeployer {
+	return &serviceDeployer{
+		mesg: mesg,
+	}
+}
+
+func (d *serviceDeployer) FromGitURL(url string) (*service.Service, *importer.ValidationError, error) {
+	d.sendStatus("Downloading service...")
+	path, err := d.createTempDir()
+	if err != nil {
+		return nil, nil, err
+	}
+	if err := d.gitClone(url, path); err != nil {
+		return nil, nil, err
+	}
+	return d.deploy(path)
+}
+
+func (d *serviceDeployer) gitClone(repoURL string, path string) error {
+	u, err := url.Parse(repoURL)
+	if err != nil {
+		return err
+	}
+	if u.Scheme == "" {
+		u.Scheme = "https"
+	}
+	options := &git.CloneOptions{}
+	if u.Fragment != "" {
+		options.ReferenceName = plumbing.ReferenceName("refs/heads/" + u.Fragment)
+		u.Fragment = ""
+	}
+	options.URL = u.String()
+	_, err = git.PlainClone(path, false, options)
+	return err
+}
+
+func (d *serviceDeployer) FromGzippedTar(r io.Reader) (*service.Service, *importer.ValidationError, error) {
+	d.sendStatus("Sending service context to core daemon...")
+	path, err := d.createTempDir()
+	if err != nil {
+		return nil, nil, err
+	}
+	if err := archive.Untar(r, path, &archive.TarOptions{
+		Compression: archive.Gzip,
+	}); err != nil {
+		return nil, nil, err
+	}
+	return d.deploy(path)
+}
+
+func (d *serviceDeployer) deploy(path string) (*service.Service, *importer.ValidationError, error) {
```

Comment:
I know golint will not highlight this, but even for unexported methods/funcs it's nice to have some docs (the docs are for mainly us, not to satisfy linters :))

Reply:
done